### PR TITLE
lxd-generate: Create filtered statements from pre-existing "objects" statement.

### PR DESCRIPTION
**Pull request by masnax**
As per our discussions on irc, I've updated `lxd-generate`'s statement generation for all statements of the type `objects-by-FIELD` to first search the package for a variable declaration of the form
```go
var <entity>Objects = RegisterStmt(`SQL String`)
```
And then extract the `SQL String` from the syntax tree, and append a `WHERE` clause with the given filters, and create a new variable declaration of the form
```go
var <entity>ObjectsByField = RegisterStmt(`SQL String with WHERE clause`)
```

This will allow us to specify our own `<entity>Objects` for more complex queries, and still be able to use `lxd-generate` to create filterable statements. 

Additionally, I went ahead and created a `lxd-generate.md` under the `docs` directory with an overview of all the functionality of `lxd-generate` including how to formulate structs and how `lxd-generate` will behave based on that, as well as all of the types of generation and associated tags that `lxd-generate` can use.

I'll send up another PR soon that will add an `sql=<table>.<column>` tag to each field that will let us bypass the generator's automatic generation of table names and column names.

### CODE REVIEW

**Comment by masnax** [comment]
If we would prefer the `lxd-generate` doc to be under `lxd/db/generate` as a README, I'll move it over so it's not included among other more usage-related docs. For now I've put it under `internals`.

**Comment by tomponline** [comment]
@ru-fu @stgraber are you happy to have lxd-generate in the "internals" section of the docs, or in the docs at all (and perhaps a separate docs directory in the lxd-generate directory)? As I understand we don't really want to encourage external usage of this tool.

**Comment by ru-fu** [comment]
> @ru-fu @stgraber are you happy to have lxd-generate in the "internals" section of the docs, or in the docs at all (and perhaps a separate docs directory in the lxd-generate directory)? As I understand we don't really want to encourage external usage of this tool.

That really depends on the intention - if we don't want users to run this tool and the internal workings aren't relevant for anybody who doesn't use it, it should rather be a README in the directory that contains the tool.
If it is good for users to know that this tool exists, or to have an idea of how it work, we should have it in the docs, and the Internals section is probably a good place for now.

**Comment by masnax** [comment]
@ru-fu Thanks, that makes me think a README is best, as this will only be relevant to LXD development, and entirely irrelevant for users.

**Comment by tomponline** [comment]
> @ru-fu Thanks, that makes me think a README is best, as this will only be relevant to LXD development, and entirely irrelevant for users.

Sounds good to me.

File: lxd/db/generate/db/stmt.go [hunk 1]
```diff
@@ -87,6 +87,10 @@ func (s *Stmt) GenerateSignature(buf *file.Buffer) error {
 }
 
 func (s *Stmt) objects(buf *file.Buffer) error {
+	if strings.HasPrefix(s.kind, "objects-by") {
```
Comment (tomponline) [inline lxd/db/generate/db/stmt.go:90]: Can you add a comment here explaining what its doing please?

File: lxd/db/generate/db/stmt.go [hunk 2]
```diff
@@ -260,6 +225,58 @@ func (s *Stmt) objects(buf *file.Buffer) error {
 	return nil
 }
 
+func (s *Stmt) objectsBy(buf *file.Buffer) error {
```
Comment (tomponline) [inline lxd/db/generate/db/stmt.go:233]: This could do with a function comment explaining what it does.

File: lxd/db/cluster/certificate_projects.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ package cluster
 //go:generate -command mapper lxd-generate db mapper -t certificate_projects.mapper.go
 //go:generate mapper reset -i -b "//go:build linux && cgo && !agent"
 //
+//go:generate mapper stmt -e certificate_project objects
```
Comment (tomponline) [inline lxd/db/cluster/certificate_projects.go:10]: What does "Add objects generation for..." mean and why is this only being applied to instance_profiles and certificate_projects

File: lxd/db/generate/db/stmt.go [hunk 4]
```diff
@@ -135,7 +135,12 @@ func (s *Stmt) objects(buf *file.Buffer) error {
 	for _, field := range fields {
 		if field.Config.Get("order") != "" {
 			if field.IsScalar() {
-				orderBy = append(orderBy, lex.Plural(lex.Snake(field.Name))+".id")
+				table, _, err := field.ScalarTableColumn()
```
Comment (tomponline) [inline lxd/db/generate/db/stmt.go:138]: Is this an unrelated bug fix? If so can it be in a separate PR please?

File: lxd/db/generate/README.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+```go
+//go:generate mapper stmt -e instance objects
+//go:generate mapper stmt -e instance objects-by-Name-and-Project
+//go:generate mapper -e instance create
+//go:generate mapper -e instance update
+//go:generate mapper -e instance delete-by-Name-and-Project
+```
+
+`//go:generate mapper stmt -e instance objects table=table_name`
```
Comment (tomponline) [inline lxd/db/generate/README.md:53]: Does this need a sub-section called Arguments or similar as it feels a bit "bolted on" at the end?

Comment (tomponline) [reply lxd/db/generate/README.md:53]: And does this apply to all of the SQL statement generation commands above or just a subset of them?

File: lxd/db/generate/README.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+There are several tags that can be added to fields of a struct that will be parsed by the `ast` package.
+
+Tag                         | Description
+:--                         | :----
+`coalesce=<value>`          | Generates a SQL coalesce function with the given value `coalesce(<column>, value)`.
+`order=yes`                 | Override the default `ORDER BY` columns with all fields specifying this tag.
+`join=<joinTable.column>`   | Applies a `JOIN` of the form `JOIN <joinTable> ON <table>.<joinTable_id> = <joinTable.id>`.
+`leftjoin=<table.column>`   | Applies a `LEFT JOIN` of the same form as a `JOIN`.
+`joinon=<column>`           | Overrides the default `JOIN ON` with the given column name, replacing `<joinTable_id>` above.
```
Comment (tomponline) [inline lxd/db/generate/README.md:67]: Is the word "default" significant in this sentence (as there are sorts of JOIN ON statements that aren't default)? Or is it that you are saying this overrides the default table used with `join`?

File: lxd/db/generate/README.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+`via=<otherTable>`          | Applies a nested join when included with a `join` tag.
+`primary=yes`               | Assigns the primary key for the table. Will default to `Name` if unspecified.
```
Comment (tomponline) [inline lxd/db/generate/README.md:69]: This is one that regularly confuses me about the DB generator. As the DB generator doesn't generate the actual schema and thus has no control over the primary key. Instead it seems to influence the minimal set of arguments that should be supplied to specify a single row (but this doesn't have to be the primary key)? Is that correct?

**Review by tomponline — CHANGES_REQUESTED** [review]
A great start on the docs,very helpful. There's a few clarifications and improvements we could make though. Thanks

File: lxd/db/generate/db/stmt.go (hunk 1 above)
Comment (masnax) [reply lxd/db/generate/db/stmt.go:90]: Sorry, the change there is basically what the first half of PR description is. I've added a commit message to describe it better.

File: lxd/db/cluster/certificate_projects.go (hunk 3 above)
Comment (masnax) [reply lxd/db/cluster/certificate_projects.go:10]: So the primary change here is that the `objects-by` directives now parse the variable declaration that's produced from the `objects` directive (or an equivalent variable declaration created manually). These two entities didn't have an `objects` directive, so their `objects-by` directives would have failed.

File: lxd/db/generate/db/stmt.go (hunk 4 above)
Comment (masnax) [reply lxd/db/generate/db/stmt.go]: To be honest, this should be taken care of by the next PR I have coming up so I've just removed it outright for now.

File: lxd/db/generate/README.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+This will initiate a call to `lxd-generate db mapper -t instances.mapper.go` on
+each call to the go generation directive `mapper`.
+
+### SQL Statement generation
+
+SQL generation supports the following SQL statement types:
+Type                                  | Description
+:---                                  | :----
+`objects`                             | Creates a basic SELECT statement of the form `SELECT <columns> FROM <table> ORDER BY <columns>`.
+`objects-by-<FIELD>-and-<FIELD>...`   | Parses a pre-existing SELECT statement of the form produced by`objects`, and appends a `WHERE` clause with the given fields located in the associated struct.
```
[the remark this answers is not in the document]
Comment (masnax) [reply lxd/db/generate/README.md]: So it should be exactly of the form produced by the `objects` directive, so that is 
```go
var <entity>Objects = <database>.RegisterStmt(`SQL String does here`)
```

I've added this to the description as well now.

File: lxd/db/generate/README.md (hunk 5 above)
Comment (masnax) [reply lxd/db/generate/README.md]: Sure, there are some other generation directive config keys that I've given their own section now. They reference the `EntityTable`, `ReferenceTable`, `MapTable`, and `AssociationTable` definitions below.

Also yes, the `table=<name>` config key applies to all `stmt` commands.

File: lxd/db/generate/README.md (hunk 6 above)
Comment (masnax) [reply lxd/db/generate/README.md]: So "default" throughout the README consistently refers to the default behaviour of `lxd-generate` without any tags present. 

So as per the description, the `join` tag will produce a clause of the form `JOIN <joinTable> ON <table>.<joinTable_id> = <joinTable.id>` and the `joinon` tag will override the `<table>.<joinTable_id>` segment of that clause.

File: lxd/db/generate/README.md (hunk 7 above)
Comment (masnax) [reply lxd/db/generate/README.md]: Yes, that's a much better description than just "assigns the primary key", I've now switched it to use that description.

File: lxd/db/generate/db/stmt.go [hunk 9]
```diff
@@ -260,6 +225,61 @@ func (s *Stmt) objects(buf *file.Buffer) error {
 	return nil
 }
 
+// objectsBy parses the variable declaration produced by the 'objects' function, and appends a WHERE clause to its SQL
+// string using the objects-by-<FIELD> field suffixes, and then creates a new variable declaration.
+// Strictly, it will look for variables of the form 'var = <entity>Objects = <database>.RegisterStmt(`SQL String`)'.
```
Comment (tomponline) [inline lxd/db/generate/db/stmt.go:230]: Did you mean `var <entity>Objects = <database>.RegisterStmt(`SQL String`)` here (comparing to the commit comment itself)?

Comment (tomponline) [reply lxd/db/generate/db/stmt.go:230]: Also does this var need to be exported? If so would be worth indicating that. But im not sure if the DB generator worries about such things :)

**Review by tomponline — APPROVED** [review]

File: lxd/db/generate/db/stmt.go (hunk 9 above)
Comment (masnax) [reply lxd/db/generate/db/stmt.go:230]: Oops, good catch! 

And no, the var doesn't need to be exported, even if it's in a different package.

**Comment by tomponline** [comment]
jenkins: test this please

File: lxd/db/generate/db/stmt.go (hunk 9 above)
Comment (masnax) [reply lxd/db/generate/db/stmt.go]: Fixed the comment.

**Comment by stgraber** [comment]
jenkins: test this please
